### ISSUE TITLE
Empty tags file without --options parameter

### ISSUE THREAD

**Issue by mgoral** [ludovicchabant/vim-gutentags#262]
Hi,
as title says, Gutentags generates empty tags file when no `--options` is set. Below logs/examples are generated when trying to auto-generate tags for Gutentags repository itself:

```
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE	u-ctags	/u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR	Universal Ctags Team	//
!_TAG_PROGRAM_NAME	Universal Ctags	/Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL	https://ctags.io/	/official site/
!_TAG_PROGRAM_VERSION	0.0.0	/1258adfc/
```

The problem seems to be, escaping of `$CTAGS_OPT_FILE` variable. Here's the output from Gutentags' trace:

```
gutentags: [job stdout]: ['Locking tags file...', 'Running file list command', 'eval git ls-files > "tags.files"', 'Ru
nning ctags on whole project', 'ctags -f "tags.temp" ""  --exclude=@/tmp/nvimYzLM5S/2 -L "tags.files"', 'Replacing tag
s file', 'mv -f "tags.temp" "tags"', 'Unlocking tags file...', 'Done.', '']
gutentags: [job stderr]: ['ctags: Warning: cannot open input file "" : No such file or directory', '']
gutentags: [job stderr]: ['']
gutentags: [job stdout]: ['']
gutentags: Finished ctags job.
```

Now, if I apply the following patch, tags are correctly generated once again:

```diff
diff --git a/plat/unix/update_tags.sh b/plat/unix/update_tags.sh
index 67609e5..40fdfd7 100755
--- a/plat/unix/update_tags.sh
+++ b/plat/unix/update_tags.sh
@@ -128,7 +128,7 @@ if [ $INDEX_WHOLE_PROJECT -eq 1 ]; then

     echo "Running ctags on whole project"
     echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
-    "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS "$CTAGS_ARG_LAST"
+    "$CTAGS_EXE" -f "$TAGS_FILE.temp" $CTAGS_OPT_FILE $CTAGS_ARGS "$CTAGS_ARG_LAST"
 else
     echo "Running ctags on \"$UPDATED_SOURCE\""
     echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
```

All I did was removing the quotes, but it's unacceptable of course. So I guess mabe adding `$CTAGS_OPT_FILE` to the command should be optional? Straightforward fix would be something like this, but I'm not sure if it's the best approach (maybe it can be simplified with building a ctags command option-by-option, variable-by-variable, and then doing `eval` on that, but I always get escaping in eval statements wrong):

```diff
diff --git a/plat/unix/update_tags.sh b/plat/unix/update_tags.sh
index 67609e5..8db4e86 100755
--- a/plat/unix/update_tags.sh
+++ b/plat/unix/update_tags.sh
@@ -127,12 +127,22 @@ if [ $INDEX_WHOLE_PROJECT -eq 1 ]; then
     fi
 
     echo "Running ctags on whole project"
-    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
-    "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS "$CTAGS_ARG_LAST"
+    if [ -z "$CTAGS_OPT_FILE" ]; then
+        echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
+        "$CTAGS_EXE" -f "$TAGS_FILE.temp" $CTAGS_ARGS "$CTAGS_ARG_LAST"
+    else
+        echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
+        "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS "$CTAGS_ARG_LAST"
+    fi
 else
     echo "Running ctags on \"$UPDATED_SOURCE\""
-    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
-    "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS --append "$UPDATED_SOURCE"
+    if [ -z "$CTAGS_OPT_FILE" ]; then
+        echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
+        "$CTAGS_EXE" -f "$TAGS_FILE.temp" $CTAGS_ARGS --append "$UPDATED_SOURCE"
+    else
+        echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
+        "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS --append "$UPDATED_SOURCE"
+    fi
 fi
 
 if [ "$POST_PROCESS_CMD" != "" ]; then
```

I'm using Debian, which uses Dash for /bin/sh, but Bash interprets these strings the same (i.e. leaves them intact). I'm running Universal Ctags on my system as well and I'm not sure if this behavior is different for other Ctags flavors (if they parse empty arguments differently).

**Comment by mgoral** [comment]
I updated gutentags trace, because initially I pasted the one from the version with my fix applied.

**Comment by swalladge** [comment]
Can confirm - I have the same issue. @mgoral's patches above fix it for me. Alternatively, something like this would work (and be consistent with the rest of how CTAGS_ARGS is built:

```patch
diff --git a/plat/unix/update_tags.sh b/plat/unix/update_tags.sh
index 67609e5..9fb893f 100755
--- a/plat/unix/update_tags.sh
+++ b/plat/unix/update_tags.sh
@@ -5,7 +5,6 @@ set -e
 PROG_NAME=$0
 CTAGS_EXE=ctags
 CTAGS_ARGS=
-CTAGS_OPT_FILE=
 TAGS_FILE=tags
 PROJECT_ROOT=
 LOG_FILE=
@@ -70,7 +69,7 @@ while getopts "h?e:x:t:p:l:L:s:o:O:P:cA" opt; do
             PAUSE_BEFORE_EXIT=1
             ;;
         o)
-            CTAGS_OPT_FILE="--options=$OPTARG"
+            CTAGS_ARGS="$CTAGS_ARGS --options=$OPTARG"
             ;;
         O)
             CTAGS_ARGS="$CTAGS_ARGS $OPTARG"
@@ -127,12 +126,12 @@ if [ $INDEX_WHOLE_PROJECT -eq 1 ]; then
     fi
 
     echo "Running ctags on whole project"
-    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
-    "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS "$CTAGS_ARG_LAST"
+    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" $CTAGS_ARGS \"$CTAGS_ARG_LAST\""
+    "$CTAGS_EXE" -f "$TAGS_FILE.temp" $CTAGS_ARGS "$CTAGS_ARG_LAST"
 else
     echo "Running ctags on \"$UPDATED_SOURCE\""
-    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" \"$CTAGS_OPT_FILE\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
-    "$CTAGS_EXE" -f "$TAGS_FILE.temp" "$CTAGS_OPT_FILE" $CTAGS_ARGS --append "$UPDATED_SOURCE"
+    echo "$CTAGS_EXE -f \"$TAGS_FILE.temp\" $CTAGS_ARGS --append \"$UPDATED_SOURCE\""
+    "$CTAGS_EXE" -f "$TAGS_FILE.temp" $CTAGS_ARGS --append "$UPDATED_SOURCE"
 fi
 
 if [ "$POST_PROCESS_CMD" != "" ]; then
```

Note that the only patch that correctly handles spaces in the opt file filename at the moment is @mgoral's second patch. 

Also discovered that the '-x <exclude pattern>' arg currently doesn't correctly handle spaces in the pattern

**Comment by mgoral** [comment]
Another alternative is to switch to e.g. Python with those scripts. `subprocess.run()` accepts a list of arguments, which is then passed to low-level `execv()` function. Python list can be easily created and we don't have to worry about escaping, which is problematic as it seems from the number of related issues on github.

```
ctags_exe = 'ctags'
out_file = 'tags.temp'
opt_file = ''
additional_args = ['arg1', 'arg2']
cmd = [ctags_exe, '-f', out_file]
if opt_file:
  cmd.extend(['--options', opt_file])
cmd.extend(additional_args)

print(' '.join(cmd))
supbrocess.run(cmd)
```

This approach however raises portability problems. Which Python version should be used? Python is generally available "everywhere", but AFAIK it's not required by POSIX and it's not a part of default installation on many systems.

**Comment by gpanders** [comment]
Noticed this today as well. I'm using a custom list files command (`git ls-files`) so the `-o` flag [is not passed to the `update_tags.sh` script](https://github.com/ludovicchabant/vim-gutentags/blob/master/autoload/gutentags/ctags.vim#L166), which causes this error. I'm seeing the `gutentags: ctags job failed, returned: 1` every time I open a file in a git repo.

Reverting to eb9e57f (the last working commit) until this is resolved.

**Comment by shagohead** [comment]
Have same issue on Mac OS Catalina & Universal Ctags

**Comment by ludovicchabant** [comment]
Right, yeah, sorry about that... should be fixed with 96cd7dbfe324e8983da68468085967e0b85dc898
